juju/juju
# vsphere: Allow specifying datastore with the root-disk-source constraint

## Description of change

Adds a new constraint root-disk-source. This will be the name of storage where the root disk for a machine should reside - it's currently only interpreted as the name of the datastore that shoud be used in vsphere. This allows units of different applications to be allocated in different datastores which is important for deployments of complex systems.

RootDiskSource is also added to instance.HardwareCharacteristics and machine instance data, so the selected datastore is stored for the created machine and can be used to filter machines to find one that matches the constraint. The new instance data and constraint attributes are included in model migrations.

## QA steps

* Bootstrap a vsphere controller.
* Deploy an application including the new constraint `--constraints="root-disk-source=datastore-name"`
* Verify that the new machine's root disk is created in the selected datastore
* Running `juju show-machine <new machine>` shows the root-disk-source in constraints and the hardware.
* Create a new machine without using the constraint.
* The hardware root-disk-source attribute is set to the default datastore for the new machine.

## Documentation changes

* The new root-disk-source constraint  and how it integrates with vSphere will need documentation.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1807960
